src/Datadog.Trace/Logging/RateLimitedLogger.cs

```diff
@@ -0,0 +1,79 @@
+using System;
+using System.Collections.Concurrent;
+using Datadog.Trace.Vendors.Serilog.Capturing;
+using Datadog.Trace.Vendors.Serilog.Core;
+using Datadog.Trace.Vendors.Serilog.Events;
+
+namespace Datadog.Trace.Logging
+{
+    internal class RateLimitedLogger : Datadog.Trace.Vendors.Serilog.Core.Logger
+    {
+        private const string LogIntervalKey = "DD_LOGGING_RATE";
+
+        private readonly ConcurrentDictionary<string, int> _logBuckets = new ConcurrentDictionary<string, int>();
+        private DateTime _intervalStarted = DateTime.UtcNow;
+        private int _rateLimitInterval = 30;
+
+        internal RateLimitedLogger(MessageTemplateProcessor messageTemplateProcessor, LogEventLevel minimumLevel, ILogEventSink sink, ILogEventEnricher enricher, Action dispose = null, LevelOverrideMap overrideMap = null)
+            : base(messageTemplateProcessor, minimumLevel, sink, enricher, dispose, overrideMap)
+        {
+            SetLoggingRate();
+        }
+
+        internal RateLimitedLogger(MessageTemplateProcessor messageTemplateProcessor, LoggingLevelSwitch levelSwitch, ILogEventSink sink, ILogEventEnricher enricher, Action dispose = null, LevelOverrideMap overrideMap = null)
+            : base(messageTemplateProcessor, levelSwitch, sink, enricher, dispose, overrideMap)
+        {
+            SetLoggingRate();
+        }
+
+        public override void Write(LogEventLevel level, Exception exception, string messageTemplate, params object[] propertyValues)
+        {
+            // If debug enabled, ALWAYS log
+            if (Tracer.Instance?.Settings.DebugEnabled != true)
+            {
+                var now = DateTime.UtcNow;
+                if (_intervalStarted.AddSeconds(_rateLimitInterval) >= now)
+                {
+                    // Clear the buckets and reset the time
+                    _intervalStarted = now;
+                    _logBuckets.Clear();
+                }
+                else if (ShouldSkip(level, exception, messageTemplate, propertyValues))
+                {
+                    // Rate limiting has been triggered
+                    return;
+                }
+            }
+
+            base.Write(level, exception, messageTemplate, propertyValues);
```

Comment:
If we get here, we're either in debug mode or this type of message is new for this time frame.